src/service/layers-impl.js

```diff
@@ -0,0 +1,634 @@
+    this.size_ = SizeWh(0, 0);
+    this.position_ = PositionLt(0, 0);
+
+    this.isLayer_ = false;
+    this.isRootLayer_ = false;
+    this.needsScrollRemeasure_ = false;
+    this.scrollLeft_ = 0;
+    this.scrollTop_ = 0;
+
+    this.children_ = [];
+  }
+
+  /**
+   * Gets the element's LayoutElement instance.
+   * @param {!Element}
+   * @return {!LayoutElement}
+   */
+  static for(element) {
+    return dev().assert(LayoutElement.forOptional(element));
+  }
+
+  /**
+   * Gets the element's LayoutElement instance, if the element has one.
+   * @param {!Element}
+   * @return {?LayoutElement}
+   */
+  static forOptional(element) {
+    return element[LAYOUT_PROP];
+  }
+
+  /**
+   * Finds the element's parent layer
+   * If the element is itself a layer, it returns the layer's parent layer.
+   * @param {!Element}
+   * @param {opt_force=} Whether to force a re-lookup
+   * @return {?LayoutElement}
+   */
+  static getParentLayer(element, opt_force) {
+    if (!opt_force) {
+      const layout = LayoutElement.forOptional(element);
+      if (layout) {
+        return layout.getParentLayer();
+      }
+    }
+
+    const win = element.ownerDocument.defaultView;
+    if (computedStyle(win, element).position == 'fixed') {
+      LayoutLayers.declareLayer(element);
+      return null;
+    }
+
+    let last = element;
+    let op = element.offsetParent;
+    for (let el = last.parentNode; el; el = el.parentNode) {
+      const layout = LayoutElement.forOptional(el);
+      if (layout && layout.isLayer()) {
+        return layout;
+      }
+
+      // now check to see if offsetParent is a fixed layer
+      if (el === op) {
+        if (computedStyle(win, op).position == 'fixed') {
+          LayoutLayers.declareLayer(op);
+          return op;
+        }
+        op = op.offsetParent;
+      }
+
+      last = el;
+    }
+
+
+    if (last.nodeType !== Node.DOCUMENT_NODE) {
+      throw dev().createError('element is not in the DOM tree');
+    }
+    return null;
+  }
+
+  contains(element) {
+    const root = this.element_;
+    return root !== element && root.contains(element);
+  }
+
+  add(child) {
+    dev().assert(this.isLayer());
+    dev().assert(this.contains(child));
+
+    const layout = LayoutElement.for(child);
+    this.children_.push(layout);
+  }
+
+  remove(child) {
+    dev().assert(this.isLayer());
+
+    const layout = LayoutElement.for(child);
+    dev().assert(layout.getParentLayer() === this);
+
+    layout.undeclareLayer();
+    layout.forgetParentLayer();
+    const i = this.children_.indexOf(layout);
+    if (i > -1) {
+      this.children_.splice(i, 1);
+    }
+  }
+
+  isLayer() {
+    return this.isLayer_;
+  }
+
+  declareLayer() {
+    this.isLayer_ = true;
+    this.needsRemeasure_ = true;
+    this.needsScrollRemeasure_ = true;
+
+    const element = this.element_;
+    const scroller = getScrollingElement(element);
+    this.isRootLayer_ = element === scroller || !scroller.contains(element);
```

Comment:
What about `position:fixed`? It's technically inside the scroller, but is outside of its layout...

Reply:
Cool. These details should really be in docs/comments.